content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+
+### 1st main bug
+
+File copy quickly hangs when copying a file.
+{{< video src="/img/blogs/hbelusca_gui-setup/guisetup_install_filecopy_hang.mp4" type="video/mp4" caption="Setting up the installation, and file copy hang">}}
+
+It turns out that this happens for every file present in the compressed `reactos.cab` installation archive.
+
+The reason of this hang is due to the fact, that when extracting compressed files, a temporary file is created _somewhere_. But where exactly?
+Looking at the code gives some clues:
+```c
+//
+// setupapi/queue.c
+//
+1012 static BOOL do_file_copyW( LPCWSTR source, LPCWSTR target, DWORD style,
+1013                            PSP_FILE_CALLBACK_W handler, PVOID context )
+1014 {
+...
+...
+1028 #ifdef __REACTOS__
+1029     /* Get a temp file name */
+1030     if (!GetTempPathW(ARRAYSIZE(TempPath), TempPath))
+1031     {
+1032         ERR("GetTempPathW error\n");
+1033         return FALSE;
+1034     }
+1035
+1036     /* Try to open the source file */
+1037     hSource = LZOpenFileW((LPWSTR)source, &OfStruct, OF_READ);
+1038     if (hSource < 0)
+1039     {
+1040         TRACE("LZOpenFileW(1) error %d %s\n", (int)hSource, debugstr_w(source));
+1041         return FALSE;
+1042     }
+1043
+1044     if (!GetTempFileNameW(TempPath, L"", 0, TempFile))
+1045     {
+...
+...
+1056         return FALSE;
+1057     }
+1058
+1059     /* Extract the compressed file to a temp location */
+1060     hTemp = LZOpenFileW(TempFile, &OfStruct, OF_CREATE);
+1061     if (hTemp < 0)
+1062     {
+...
+...
+1076         return FALSE;
+1077     }
+1078
+1079     lRes = LZCopy(hSource, hTemp);
+1080
+```
+
+(_Important NOTE:_ This is ReactOS-specific code to support a behaviour similar to what Windows has, and which is _**NOT**_ currently supported on Wine.
+_However_, Wine code also uses a temporary destination file to perform the copy, instead of directly copying over any possible existing file. See below for the reason why.)
+
+The default `TEMP` directory for the currently-running ReactOS installation is retrieved by a call to `GetTempPathW()` -- then followed by an attempt to open the source with `LZOpenFileW()` that also serves to handle single-compressed source files.
+(Remember those `MSPAINT.EX_`-like files in your Windows &le; 2003 installation disks? Those are compressed single-file CAB archives, containing one single file, in this example: `MSPAINT.EXE`.)
+This call will succeed even if the file is _not_ compressed.
+
+Then, a temporary destination file path is generated with `GetTempFileNameW()`, based on the retrieved `TEMP` directory path.
+Finally the source file is extracted to this temporary file (or just copied to, in case it was not compressed).
+Only then (omitted code that follows the snippet above), the created temporary file is moved to its final target.
+(The reason why this temporary file step is necessary, is for handling copy/replacements over files that could be already existing and currently in use.)
```

Comment:
```suggestion
This temporary file step is necessary for handling copy/replace operations on already existing files that are currently in use.
```

Reply:
I have added to this sentence: `"..., for which the [operation is delayed until reboot"`